### PR TITLE
Fix link wrapping punctuation

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- add helper for punctuation detection
- adjust wrapping to keep trailing punctuation with preceding link
- test punctuation handling after link wrap

## Testing
- `make fmt`
- `make lint`
- `make test`

------
https://chatgpt.com/codex/tasks/task_e_687f3b34074c8322bef6bb579224f3f8

## Summary by Sourcery

Fix link wrapping to prevent orphaned punctuation by introducing punctuation detection and updating the wrapping logic

Bug Fixes:
- Ensure trailing punctuation remains attached to a wrapped link instead of being orphaned

Enhancements:
- Add is_punctuation helper function to detect ASCII punctuation tokens

Tests:
- Add test to verify punctuation stays with wrapped links

### CODE REVIEW

File: src/wrap.rs [hunk 1]
```diff
@@ -191,6 +191,10 @@ fn should_break_line(width: usize, current_width: usize, last_split: Option<usiz
     current_width > width && last_split.is_some()
 }
 
+const TRAILING_PUNCTUATION: &[char] = &['.', ',', ';', ':', '?', '!', '…'];
+
+fn is_punctuation(token: &str) -> bool { token.chars().all(|c| TRAILING_PUNCTUATION.contains(&c)) }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/wrap.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

[the remark this answers is not in the document]
Comment (leynos) [reply src/wrap.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

```
Please address the comments from this code review:
## Overall Comments
- is_punctuation currently treats any ASCII punctuation token (e.g. ‘[’ or ‘]’) as a break—consider narrowing that to just the typical trailing punctuation (.,;:?!…) or using Unicode punctuation categories so you don’t misclassify markdown delimiters.
- Rather than a separate special‐case block for punctuation after wrapping, you could fold punctuation detection into your existing last_split logic to streamline the flow and avoid extra continues.

## Individual Comments

### Comment 1
<location> `tests/wrap.rs:478` </location>
<code_context>
     );
 }
+
+/// Ensures that punctuation following a wrapped link is not orphaned.
+#[test]
+fn test_wrap_link_trailing_punctuation() {
+    let input = lines_vec![
+        "See [HTML table support for more \
+         details](docs/architecture.md#html-table-support-in-mdtablefix).",
+    ];
+    let expected = lines_vec![
+        "See",
+        "[HTML table support for more \
+         details](docs/architecture.md#html-table-support-in-mdtablefix).",
+    ];
+    let output = process_stream(&input);
+    assert_eq!(output, expected);
+}
</code_context>

<issue_to_address>
Consider adding tests for multiple and leading punctuation cases.

Please add tests for multiple trailing punctuation marks, various punctuation types, and punctuation before the link to ensure comprehensive coverage.

Suggested implementation:

```rust
/// Ensures that punctuation following a wrapped link is not orphaned.
#[test]
fn test_wrap_link_trailing_punctuation() {
    let input = lines_vec![
        "See [HTML table support for more \
         details](docs/architecture.md#html-table-support-in-mdtablefix).",
    ];
    let expected = lines_vec![
        "See",

```

```rust
        "[HTML table support for more \
         details](docs/architecture.md#html-table-support-in-mdtablefix).",
    ];
    let output = process_stream(&input);
    assert_eq!(output, expected);
}

/// Ensures that multiple trailing punctuation marks after a wrapped link are not orphaned.
#[test]
fn test_wrap_link_multiple_trailing_punctuation() {
    let input = lines_vec![
        "Check this [link](foo.md)!?",
    ];
    let expected = lines_vec![
        "Check this",
        "[link](foo.md)!?",
    ];
    let output = process_stream(&input);
    assert_eq!(output, expected);
}

/// Ensures that various punctuation types after a wrapped link are not orphaned.
#[test]
fn test_wrap_link_various_trailing_punctuation() {
    let input = lines_vec![
        "Reference [doc](bar.md):",
        "See [note](baz.md)...",
        "Alert [warn](warn.md);",
    ];
    let expected = lines_vec![
        "Reference",
        "[doc](bar.md):",
        "See",
        "[note](baz.md)...",
        "Alert",
        "[warn](warn.md);",
    ];
    let output = process_stream(&input);
    assert_eq!(output, expected);
}

/// Ensures that punctuation before a link is handled correctly.
#[test]
fn test_wrap_link_leading_punctuation() {
    let input = lines_vec![
        "\"[Quoted link](quote.md)\" is important.",
        "([Parenthesized link](paren.md)) is here.",
    ];
    let expected = lines_vec![
        "\"",
        "[Quoted link](quote.md)\" is important.",
        "([Parenthesized link](paren.md)) is here.",
    ];
    let output = process_stream(&input);
    assert_eq!(output, expected);
}

/// Ensures that both leading and trailing punctuation around a link are handled.
#[test]
fn test_wrap_link_leading_and_trailing_punctuation() {
    let input = lines_vec![
        "\"[Link](foo.md)!\"",
        "([Another](bar.md)?)",
    ];
    let expected = lines_vec![
        "\"",
        "[Link](foo.md)!\"",
        "([Another](bar.md)?)",
    ];
    let output = process_stream(&input);
    assert_eq!(output, expected);
}

```
</issue_to_address>
```

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### src/wrap/tests.rs (1)

`1-511`: _⚠️ Potential issue_ | _🟠 Major_ | _🏗️ Heavy lift_
**File exceeds maximum line limit.**
This file is 511 lines long, exceeding the 400-line limit mandated by the coding guidelines by a significant margin (111 lines). Decompose the test module into separate focused test modules or move large test fixtures to external files.
Triage: `[type:syntax/md]`
As per coding guidelines: Files must not exceed 400 lines in length. Large modules must be decomposed.

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/wrap/tests.rs` around lines 1 - 511, This test file exceeds the 400-line
limit; split it into smaller test modules by grouping related tests and moving
large fixtures out: extract the TRAILING_PUNCTUATION_CHARS constant and all
proptest cases/wrapping tests (e.g.,
wrap_preserving_code_keeps_supported_punctuation_with_links,
wrap_preserving_code_keeps_supported_punctuation_with_code_spans, and related
proptest blocks) into a new tests module focused on punctuation and wrapping,
move wrap_text-focused tests (e.g., wrap_text_preserves_hyphenated_words,
wrap_text_preserves_code_spans, wrap_text_multiple_code_spans,
wrap_text_nested_backticks, wrap_text_unmatched_backticks,
wrap_text_preserves_links,
wrap_text_keeps_trailing_spaces_for_blockquote_final_line,
wrap_text_keeps_trailing_spaces_for_bullet_final_line,
wrap_text_preserves_indented_hash_as_text, wrap_text_flushes_before_heading)
into another module, and place small unit tests around tokenization
(determine_token_span_groups_related_tokens,
segment_inline_splits_before_embedded_links,
segment_inline_preserves_escaped_link_literals,
attach_punctuation_appends_to_previous_code_line,
attach_punctuation_requires_empty_current_buffer,
attach_punctuation_ignores_non_code_suffix) together; ensure you update mod
declarations (fence_tracker, prefix) and any use/imports so each new file
imports super::{inline::{attach_punctuation_to_previous_line,
determine_token_span, wrap_preserving_code}, tokenize::segment_inline} and
crate::wrap::{BlockKind, classify_block, wrap_text} as needed and keep test
names unchanged to preserve intent.

#### src/wrap/inline.rs (1)

`1-416`: _⚠️ Potential issue_ | _🟠 Major_ | _🏗️ Heavy lift_
**File exceeds maximum line limit.**
This file is 416 lines long, exceeding the 400-line limit mandated by the coding guidelines. Decompose large modules or move inline data/fixtures to external files.
Triage: `[type:syntax/md]`
As per coding guidelines: Files must not exceed 400 lines in length. Large modules must be decomposed.

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/wrap/inline.rs` around lines 1 - 416, This file exceeds the 400-line
limit—split it into smaller modules: extract postprocessing logic and tests plus
large helper groups into new files and re-export as needed. Specifically, move
the postprocess module (postprocess::{merge_whitespace_only_lines,
rebalance_atomic_tails}) into its own module file and relocate
test_support/tests into separate test modules; extract the fragment-related
types and helpers (InlineFragment, FragmentKind, classify_fragment,
width_as_f64, push_span_text, build_fragments, and related impls) into a new
inline_fragments module and keep only the wrapping orchestration
(determine_token_span, merge_code_span, extend_punctuation,
should_couple_whitespace, is_trailing_punct, is_trailing_punctuation_token,
looks_like_link, is_whitespace_token, is_inline_code_token, and
wrap_preserving_code) in the current file, updating imports and pub(super)
re-exports so callers of wrap_preserving_code still compile.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Please suggest fixes for the following concerns and provide an AI coding agent prompts for the fixes:

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Unit And Behavioural) | ⚠️ Warning | Tests prioritise internal helper testing via pub(super) and pub(in crate) functions over public API coverage. No tests verify link integrity under edge cases like malformed links or split boundaries. | Add end-to-end tests invoking process\_stream with edge-case links: empty URL `[]()`, unbalanced parens, nested delimiters, to verify links are never split mid-token during wrapping. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Link wrapping tests use unit assertions instead of snapshot tests, whilst similar wrapping features (inline code, link references) use insta snapshots for text-based output. | Add snapshot tests using insta for link punctuation wrapping scenarios, covering long links with trailing punctuation, multiple trailing marks, various punctuation types, and leading punctuation cases. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | The `test_wrap_link_at_exact_wrap_boundary_is_not_split` test is vacuous; it only asserts the link exists somewhere in output, not that it's at the boundary or tests punctuation handling. | Use WRAP\_COLS=80 to calculate prefix length, ensuring the link starts exactly at the wrap boundary, and verify link+punctuation remains together at that edge. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests exercise behaviour substantively. Two issues: British spelling "finalised" (not "finalized") at line 241 of src/wrap/inline.rs; wrap boundary test hardcodes 80 instead of WRAP\_COLS. | Change "finalised" to "finalized" at line 241 of src/wrap/inline.rs. Tie wrap boundary test to production WRAP\_COLS constant instead of hardcoded value. |
| Rust Compiler Lint Integrity | ❌ Error | Line 241 in src/wrap/inline.rs retains non-Oxford spelling "finalised" instead of required "finalized". No lint suppressions or dead code detected. | Replace "finalised" with "finalized" on line 241 of src/wrap/inline.rs to comply with en-GB-oxendict spelling requirements. |
| Architectural Complexity And Maintainability | ⚠️ Warning | Boundary test hard-codes wrap width (80) instead of sourcing production WRAP\_COLS constant, risking silent test drift if wrap width changes. | Expose WRAP\_COLS via a public test accessor or use a test-scoped re-export, then replace the hard-coded constant with a source-of-truth reference. |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Rust Compiler Lint Integrity | ❌ Error | span.clone() on Range at src/wrap/inline.rs:197 is unnecessary; Range implements Copy and clones introduced during implementation must be revisited before merge. | Remove .clone() from tokens\[span.clone()\] at line 197; use tokens\[span\] directly since Range is Copy and can be reused. |
| Developer Documentation | ⚠️ Warning | PR exposed `WRAP_COLS` as public API but failed to document this change in developers-guide.md. Only internal `is_trailing_punctuation_token` helper was documented. | Document `WRAP_COLS` constant in docs/developers-guide.md explaining it is the public column-width constant available for test integration. |
| Testing (Property / Proof) | ⚠️ Warning | The proptest for punctuation-link attachment uses fixed wrap width (24) and fixed link format. Comparable property tests vary wrap widths; this is essential for layout-dependent wrapping invariants. | Extend the proptest to vary wrap widths (e.g., 24usize..96) and generate arbitrary link URLs via a strategy, matching the coverage pattern of wrap\_text\_keeps\_generated\_footnote\_references\_atomic. |
